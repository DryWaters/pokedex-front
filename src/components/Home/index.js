import React, { Component } from 'react';
import Header from '../Header';
import PokemonCard from '../PokemonCard';
import LocalJSON from '../../local_data/example_single.json';


class Home extends Component {

  constructor() {
    super();

  }

  componentDidMount() {
    this.fetchPokemon()
  }

  fetchPokemon() {

  } 

  render() {
    return (
      <div>
        <Header />
        <PokemonCard {...LocalJSON} />
      </div>
    )
  }
}

export default Home;