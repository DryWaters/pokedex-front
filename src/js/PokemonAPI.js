const baseURL = 'https://pokedex-backend-server.herokuapp.com';

export const getRangeOfPokemon = (id, range) =>
  fetch(`${baseURL}/pokemon?id=${id}&range=${range}`)
    .then(res => res.json())
    .then(json => json.pokemon);

export const getPokemonSprite = (image_path) => baseURL + image_path;
