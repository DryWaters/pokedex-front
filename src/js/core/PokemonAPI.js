const baseURL = 'https://pokedex-backend-server.herokuapp.com';

export const getRangeOfPokemon = (id, range) =>
  fetch(`${baseURL}/pokemon?id=${id}&range=${range}`)
    .then(res => res.json())
    .then(json => json.pokemon);

export const getPokemonDetails = (id) =>
  fetch(`${baseURL}/pokemon/${id}`)
    .then(res => res.json());

export const getPokemonSprite = (image_path) => baseURL + image_path;
