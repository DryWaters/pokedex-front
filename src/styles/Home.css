.card-container {
  width: 100%;
  max-width: 1300px; /* 4 cards with some spacing */
  justify-content: space-around;
  display: flex;
  flex-wrap: wrap;
  padding: 20px;
}
