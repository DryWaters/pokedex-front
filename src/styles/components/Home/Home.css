.cardContainer {
  width: 100%;
  justify-content: space-between;
  /* align-content: space-around; */
  display: flex;
  flex-wrap: wrap;
  padding: 20px;
}