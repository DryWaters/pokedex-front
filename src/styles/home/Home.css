.main {
  width: 100%;
  height: 100%;
}

.card-container {
  width: 100%;
  max-width: 1300px; /* 4 cards with some spacing */
  justify-content: space-around;
  display: flex;
  flex-wrap: wrap;
  padding: 20px;
}

.load-button {
  margin: 10px;
}
