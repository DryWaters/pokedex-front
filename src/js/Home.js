import React, { Component } from 'react';

import { getRangeOfPokemon } from './PokemonAPI.js';
import Header from './Header.js';
import PokemonCard from './PokemonCard.js';
import NoPokemon from './NoPokemon.js';

import '../styles/Home.css'

class Home extends Component {

  constructor() {
    super();
    this.state = {
      pokemon: []
    }
  }

  componentDidMount() {
    this.fetchPokemon(1, 20);
  }

  fetchPokemon(id, range) {
    getRangeOfPokemon(id, range)
      .then(newPokemon => {
        const oldPokemon = this.state.pokemon.slice();
        this.setState({
          pokemon: [...oldPokemon, ...newPokemon]
        });
      })

      // TODO: Need to do something if error on fetch from server
      .catch(error => {
        window.console.log('Unable to fetch Pokeon data from server with error: ' + error);
      });
  }

  displayPokemonCards() {
    if (this.state.pokemon.length === 0) {
      return <NoPokemon />
    }
    return this.state.pokemon.map(pokemon => {
      return <PokemonCard {...pokemon} key={pokemon.id} />
    });
  }

  render() {
    return (
      <div>
        <Header />
        <div className="cardContainer">
          {this.displayPokemonCards()}
        </div>
      </div>
    )
  }
}

export default Home;
