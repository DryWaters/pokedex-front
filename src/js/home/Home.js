import React from 'react';
import InfiniteScroll from 'react-infinite-scroller';
import { Button } from 'reactstrap';

import { getRangeOfPokemon, getPokemonDetails } from '../core/PokemonAPI';
import Header from './Header';
import PokemonCard from './PokemonCard';
import Loader from './Loader';
import LoadingError from './LoadingError';
import PokemonModal from '../modal/PokemonModal';

import '../../styles/home/Home.css';

class Home extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      pokemon: [],
      pokemonDetails: null,
      error: '',
      shouldInfiniteScroll: false,
      initialLoad: true,
      modal: false,
    };
    this.perPage = 20;

    this.handlePokemonCardClick = this.handlePokemonCardClick.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }

  /**
   * Fetches a page of pokemon from the server and adds them to the current state
   * @param  {int} page - the page to fetch (1 indexed)
   * @return {void}
   */
  fetchPokemon(page) {
    if (this.state.error)
        return;

    const id = (page - 1) * this.perPage + 1;
    let range;
    if (page === 41) {
      this.setState({ shouldInfiniteScroll: false })
      range = 6
    } else {
      range = this.perPage;
    }
    getRangeOfPokemon(id, range)
      .then(newPokemon => {
        const oldPokemon = this.state.pokemon.slice();
        this.setState({
          pokemon: [...oldPokemon, ...newPokemon],
          initialLoad: false,
        });
      })

      .catch(error => {
        this.setState({
            error: error.message,
            shouldInfiniteScroll: false,
        });
      });
  }

  displayPokemonCards () {
    return this.state.pokemon.map(pokemon => {
      return <PokemonCard {...pokemon} key={pokemon.id} handlePokemonCardClick={this.handlePokemonCardClick} />
    });
  }

  renderLoadButton () {
    return (
      <div>
        <Button color="primary" size="lg" onClick={this.enableInfiniteScroll.bind(this)}> Load more Pokemon </Button>
      </div>
    );
  }

  enableInfiniteScroll (evt) {
    evt.preventDefault();
    this.setState({ shouldInfiniteScroll: true });
  }

   /**
   * Fetches a pokemon details from server and adds them to the current state
   * @param  {int} id - the id of the pokemon to fetch
   * @return {void}
   */
  handlePokemonCardClick (id) {
    if (this.state.error ||
      id < 1 ||
      id > 806)
      return;

    getPokemonDetails(id)
      .then(pokemonDetails => {
        this.setState({
          pokemonDetails,
          modal: true
        });
      })

      .catch(error => {
        this.setState({
            error: error.message,
        });
      });
  }

  handleCloseModal() {
    this.setState({modal: false})
  }

  render () {
    let cards = this.displayPokemonCards();
    let loadMoreButton = !this.state.shouldInfiniteScroll && !this.state.error ?
      this.renderLoadButton() : null;
    let error = this.state.error ? <LoadingError error={this.state.error} /> : null;

    return (
      <div className="main" align="center">
        <PokemonModal handleCloseModal={this.handleCloseModal}
          modal={this.state.modal} {...this.state.pokemonDetails} />
        <Header />
        <InfiniteScroll
          initialLoad={true}
          pageStart={0}
          loadMore={this.fetchPokemon.bind(this)}
          hasMore={this.state.shouldInfiniteScroll || this.state.initialLoad}
          loader={<Loader key="loader" />}
        >
          <div className="card-container">
            {cards}
          </div>
        </InfiniteScroll>
        {loadMoreButton}
        {error}
      </div>
    )
  }
}

export default Home;
