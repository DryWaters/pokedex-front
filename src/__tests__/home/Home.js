import React from 'react';
import { shallow } from 'enzyme';

import Home from '../../js/home/Home';

import * as PokemonAPI from '../../js/core/PokemonAPI';
import ExampleJSON from '../../../local_data/example_detail.json';

describe('Home Component', () => {

  it('should render Home component without crashing', () => {
    shallow(<Home />);
  });

  it('should match Home component snapshot', () => {
    const wrapper = shallow(<Home />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('toggle', () => {
    let wrapper, instance;
    beforeEach(() => {
      wrapper = shallow(<Home />);
      instance = wrapper.instance();
    });

    it('should set `state.modal` to false when closing modal', () => {
      wrapper.setState({ modal: true });
      expect(wrapper.state('modal')).toBe(true);

      instance.handleCloseModal();
      expect(wrapper.state('modal')).toBe(false);
    });
  })

  describe('Test API is being called', () => {
    let wrapper, instance;
    beforeEach(() => {
      wrapper = shallow(<Home />);
      instance = wrapper.instance();
    });
    it('should call API if clicked on pokemon card', () => {
      PokemonAPI.getPokemonDetails = jest.fn();
      PokemonAPI.getPokemonDetails.mockImplementation(() => Promise.resolve(ExampleJSON))

      return instance.handlePokemonCardClick(1)
        .then(() => {
          expect(PokemonAPI.getPokemonDetails).toHaveBeenCalledTimes(1);
          expect(instance.state.modal).toBe(true);
          expect(instance.state.pokemonDetails).toEqual(ExampleJSON)
        });

    });

    it('should call API when fetching pokemon', () => {
      PokemonAPI.getRangeOfPokemon = jest.fn();
      PokemonAPI.getRangeOfPokemon.mockImplementation(() => Promise.resolve([{
        "id": 1,
        "name": "Bulbasaur",
        "types": [
          "grass",
          "poison"
        ],
        "image_path": "/sprites/pokemon/large/1.png"
      }]
      ));


      return instance.fetchPokemon(1)
        .then(() => {
          expect(PokemonAPI.getRangeOfPokemon).toHaveBeenCalledTimes(1);
          expect(instance.state.pokemon).toEqual(
            expect.arrayContaining([
              expect.objectContaining({
                "id": 1,
                "name": "Bulbasaur",
                "types": [
                  "grass",
                  "poison"
                ],
                "image_path": "/sprites/pokemon/large/1.png"
              })
            ])
          )
        });
    });
  });
});
