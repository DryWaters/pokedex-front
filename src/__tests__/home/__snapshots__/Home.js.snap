// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should match Home component snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "align": "center",
      "children": Array [
        <PokemonModal
          data={null}
          getPokemonDetails={[Function]}
          handleCloseModal={[Function]}
          modal={false}
          selectForm={[Function]}
          selectedForm={0}
/>,
        <Header />,
        <Search />,
        <InfiniteScroll
          element="div"
          hasMore={true}
          initialLoad={true}
          isReverse={false}
          loadMore={[Function]}
          loader={<Loader />}
          pageStart={0}
          ref={null}
          threshold={250}
          useCapture={false}
          useWindow={true}
>
          <div
                    className="card-container"
          >
                    
          </div>
</InfiniteScroll>,
        <div>
          <Button
                    className="load-button"
                    color="primary"
                    onClick={[Function]}
                    size="lg"
                    tag="button"
          >
                    Load more Pokemon
          </Button>
</div>,
        null,
      ],
      "className": "main",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": null,
          "getPokemonDetails": [Function],
          "handleCloseModal": [Function],
          "modal": false,
          "selectForm": [Function],
          "selectedForm": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            className="card-container"
>
            
</div>,
          "element": "div",
          "hasMore": true,
          "initialLoad": true,
          "isReverse": false,
          "loadMore": [Function],
          "loader": <Loader />,
          "pageStart": 0,
          "ref": null,
          "threshold": 250,
          "useCapture": false,
          "useWindow": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "card-container",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Button
            className="load-button"
            color="primary"
            onClick={[Function]}
            size="lg"
            tag="button"
>
            Load more Pokemon
</Button>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Load more Pokemon",
            "className": "load-button",
            "color": "primary",
            "onClick": [Function],
            "size": "lg",
            "tag": "button",
          },
          "ref": null,
          "rendered": "Load more Pokemon",
          "type": [Function],
        },
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "align": "center",
        "children": Array [
          <PokemonModal
            data={null}
            getPokemonDetails={[Function]}
            handleCloseModal={[Function]}
            modal={false}
            selectForm={[Function]}
            selectedForm={0}
/>,
          <Header />,
          <Search />,
          <InfiniteScroll
            element="div"
            hasMore={true}
            initialLoad={true}
            isReverse={false}
            loadMore={[Function]}
            loader={<Loader />}
            pageStart={0}
            ref={null}
            threshold={250}
            useCapture={false}
            useWindow={true}
>
            <div
                        className="card-container"
            >
                        
            </div>
</InfiniteScroll>,
          <div>
            <Button
                        className="load-button"
                        color="primary"
                        onClick={[Function]}
                        size="lg"
                        tag="button"
            >
                        Load more Pokemon
            </Button>
</div>,
          null,
        ],
        "className": "main",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data": null,
            "getPokemonDetails": [Function],
            "handleCloseModal": [Function],
            "modal": false,
            "selectForm": [Function],
            "selectedForm": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <div
              className="card-container"
>
              
</div>,
            "element": "div",
            "hasMore": true,
            "initialLoad": true,
            "isReverse": false,
            "loadMore": [Function],
            "loader": <Loader />,
            "pageStart": 0,
            "ref": null,
            "threshold": 250,
            "useCapture": false,
            "useWindow": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "card-container",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Button
              className="load-button"
              color="primary"
              onClick={[Function]}
              size="lg"
              tag="button"
>
              Load more Pokemon
</Button>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Load more Pokemon",
              "className": "load-button",
              "color": "primary",
              "onClick": [Function],
              "size": "lg",
              "tag": "button",
            },
            "ref": null,
            "rendered": "Load more Pokemon",
            "type": [Function],
          },
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
